keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,652 @@
+
+In addition, we want to terminate running the rest of the pods if the leader index (index=0) is succeeded
+because the workers often don't have any ways to terminate themselves due to launching daemon processes like ssh-server.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successPolicy:
+  - target:
+      succeededIndexes: "0"
+      whenTargetAchieved: terminate
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./sample"]
+```
+
+#### Story 2
+
+As a simulation researcher/engineer for fluid dynamics/biochemistry.
+I want to mark the job as successful and continue running the rest of the pods if some indexes succeed
+because I set the minimum required value for sampling in the `.successPolicy.target.count` and
+perform the same simulation in all indexes.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successPolicy:
+  - target:
+      count: 5
+      whenTargetAchieved: continue
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+        - name: job-container
+          image: job-image
+          command: ["./sample"]
+```
+
+### Story 3
+
+As a simulation researcher/engineer for fluid dynamics/biochemistry.
+I want to mark the job as successful and to perform the matched `whenTargetAchieved` of `successPolicy`
+when the job meets the one of following conditions:
+
+1. Once the leader index (index=0) is succeeded, I want to terminate running the rest of the pods.
+2. Once some worker indexes (index=1+) are succeeded, I want to continue running the rest of the pods for sampling.
+
+because succeeded leader index means that the whole simulation is succeeded,
+and succeeded some worker indexes means that the minimum required value for sampling is satisfied.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successPolicy:
+  - target:
+      count: 1
+      index: 0
+      whenTargetAchieved: terminate
+  - target:
+      count: 5
+      index: "1-9"
+      whenTargetAchieved: continue
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+        - name: job-container
+          image: job-image
+          command: ["./sample"]
+```
+
+
+### Notes/Constraints/Caveats (Optional)
+
+#### No support JobSuccessPolicy for the NonIndexed Job
+
+As I described in [Non-Goals](#non-goals), we don't support the SuccessPolicy for the job with `NonIndexed` mode.
+
+#### Interchangeability between concepts of "completed" and "succeeded"
+In this KEP, we introduce the new `succeeded` concept, 
+which is interchangeable with the existing `completed` concept.
+
+It means that if the job meets the dedicated successPolicy, the `Completed` condition is added to the job.  
```

Comment:
As @kannon92 mentions here: https://github.com/kubernetes/enhancements/pull/4062#discussion_r1467776240, Giving interchangeability to `completed` and `succeeded` might cause confusion. So I will update this KEP with a new condition, `succeeded` separate from `completed`.


cc: @alculquicondor @mimowo

Reply:
It would be good to explore how the new condition is going to be used by the existing consumers and how it relates to the goals and user stories of this KEP.

Reply:
I think the condition is a good idea to mark that the successPolicy is already achieved, but the job as a whole is not yet "Completed". I'm thinking of:
```
Type: Succeeded
Status: True
Reason: SuccessPolicyTargetAchieved
Message: ...
```
Then `Completed` is only set when all pods are done (as currently). I agree though it would be good to tie this back to the user stories. I suppose one way the the new condition could be useful is to give early feedback to the users that the minimal target is already achieved, then based on their budget or time they could either continue or kill the remaining pods.

Reply:
If a Job can never change `Succeeded` from True to False (bar someone fiddling with its `.status` manually), then yes I support this.

If there's any action I can take that means the control plane indirectly changes `Succeeded` from True to False on any Job, then I think the condition needs a different name.

Reply:
> It would be good to explore how the new condition is going to be used by the existing consumers and how it relates to the goals and user stories of this KEP.

@atiratree Yes, you're right. I'm going to update this KEP.

Reply:
> If a Job can never change Succeeded from True to False (bar someone fiddling with its .status manually), then yes I support this.
> 
> If there's any action I can take that means the control plane indirectly changes Succeeded from True to False on any Job, then I think the condition needs a different name.

@sftim  That's a nice point. I think so too. I also think that the `Succeeded` condition shouldn't be updated `true` to `false`.  Thanks!

Reply:
Done.